Mathlib/NumberTheory/SiegelsLemma.lean

```diff
@@ -0,0 +1,237 @@
+
+## Notation
+
+ - `‖_‖ ` : Matrix.seminormedAddCommGroup is the sup norm, the maximum of the absolute values of
+ the entries of the matrix
+
+## References
+
+See [M. Hindry and J. Silverman, Diophantine Geometry: an Introduction][hindrysilverman00].
+-/
+
+/- We set ‖⬝‖ to be Matrix.seminormedAddCommGroup  -/
+attribute [local instance] Matrix.seminormedAddCommGroup
+
+open Matrix Finset
+
+section generalMatrices
+
+variable {m n α : Type*} [Fintype m] [Fintype n] [SeminormedAddCommGroup α]
+
+lemma exists_entry_norm_eq_norm_Matrix [Nonempty m] [Nonempty n] (A : Matrix m n α) :
+    ∃ (i : m) (j : n), ‖A‖ = ‖A i j‖₊ := by
+  simp only [norm_eq_sup_sup_nnnorm]
+  rcases Finset.exists_mem_eq_sup univ univ_nonempty (fun i => univ.sup fun j => ‖A i j‖₊)
+    with ⟨i, hi⟩
+  use i
+  rcases Finset.exists_mem_eq_sup univ univ_nonempty (fun j => ‖A i j‖₊) with ⟨j, hj⟩
+  use j
+  rw [hi.2, hj.2]
+
+end generalMatrices
+
+namespace Int.Matrix
+
+variable (m n : ℕ) (A : Matrix (Fin m) (Fin n) ℤ) (v : Fin n → ℤ) (hn : m < n)
+(hm : 0 < m)
+
+--Some definitions and relative properties
+
+local notation3 "e" => m / ((n : ℝ) - m) --exponent
+local notation3 "B" => Nat.floor (((n : ℝ) * ‖A‖) ^ e)
+-- B' is the vector with all components = B
+local notation3 "B'" => fun _ : Fin n => (B : ℤ)
+-- T is the box [0 B]^n
+local notation3 "T" =>  Finset.Icc 0 B'
+local notation3  "P" => fun i : Fin m => (∑ j : Fin n, B * posPart (A i j))
+local notation3  "N" => fun i : Fin m =>  (∑ j : Fin n, B * (- negPart (A i j)))
+-- S is the box where the image of T goes
+local notation3  "S" => Finset.Icc N P
+
+section preparation
+
+/- In order to apply Pigeohole we need:
+# Step 1: ∀ v ∈  T, (A.mulVec v) ∈  S
+and
+# Step 2: S.card < T.card
+Pigeohole will give different x and y in T with A.mulVec x = A.mulVec y in S
+Their difference is the solution we are looking for
+-/
+
+-- # Step 1: ∀ v ∈  T, (A.mulVec v) ∈  S
+
+private lemma image_T_subset_S : ∀ v ∈ T, (A.mulVec v) ∈ S := by
+  intro v hv
+  rw [mem_Icc] at hv
+  rw [mem_Icc]
+  have mulVec_def : A.mulVec v =
+      fun i => Finset.sum univ fun (j : (Fin n)) => A i j * v j := by rfl
+  rw [mulVec_def] -- unfold def of MulVec
+  refine ⟨fun i ↦ ?_, fun i ↦ ?_⟩ -- 2 goals
+  all_goals simp only [mul_neg]
+  all_goals gcongr (∑ _ : Fin n, ?_) with j _ -- get rid of sums
+  all_goals rw [← mul_comm (v j)] -- move A i j to the right of the products
+  all_goals by_cases hsign : 0 ≤ A i j   --we have to distinguish cases: we have now 4 goals
+  · rw [negPart_eq_zero.2 hsign]
+    exact mul_nonneg (hv.1 j) hsign
+  · rw [not_le] at hsign
+    rw [negPart_eq_neg.2 (le_of_lt hsign)]
+    simp only [mul_neg, neg_neg]
+    exact mul_le_mul_of_nonpos_right (hv.2 j) (le_of_lt hsign)
+  · rw [posPart_eq_self.2  hsign]
+    exact mul_le_mul_of_nonneg_right (hv.2 j) hsign
+  · rw [not_le] at hsign
+    rw [posPart_eq_zero.2 (le_of_lt hsign)]
+    exact mul_nonpos_of_nonneg_of_nonpos (hv.1 j) (le_of_lt hsign)
```

Comment:
Together wthe the `by_cases` line, this can be simplified a bit:
```lean
    rcases le_total 0 (A i j) with hsign | hsign -- we have to distinguish cases: we have now 4 goals
  · rw [negPart_eq_zero.2 hsign]
    exact mul_nonneg (hv.1 j) hsign
  · rw [negPart_eq_neg.2 hsign]
    simp only [mul_neg, neg_neg]
    exact mul_le_mul_of_nonpos_right (hv.2 j) hsign
  · rw [posPart_eq_self.2  hsign]
    exact mul_le_mul_of_nonneg_right (hv.2 j) hsign
  · rw [posPart_eq_zero.2 hsign]
    exact mul_nonpos_of_nonneg_of_nonpos (hv.1 j) hsign
```